File: rust/clock.rs
#[derive(Debug)]
pub struct Clock {
    hours: i32,
    minutes: i32,
}

impl Clock {
    pub fn new(mut hours: i32, mut minutes: i32) -> Self {
        while minutes < 0 {
            minutes += 60;
            hours -= 1;
        }
        while hours < 0 {
            hours += 24;
        }
        hours += minutes / 60;
        let c = Clock {
            hours: hours % 24,
            minutes: minutes % 60,
        };
        c
    }

    pub fn add_minutes(self, minutes: i32) -> Self {
        Clock::new(self.hours, self.minutes + minutes)
    }

    pub fn to_string(self) -> String {
        format!("{:02}:{:02}", self.hours, self.minutes)
    }
}

impl PartialEq for Clock {
    fn eq(&self, other: &Clock) -> bool {
        self.hours == other.hours && self.minutes == other.minutes
    }
}

File: lua/clock.lua
local Clock = {}
Clock.__index = Clock

function Clock.at(hours, minutes)
    minutes = minutes or 0
    local t = {
        hours = (hours + minutes // 60) % 24,
        minutes = minutes % 60,
    }
    return setmetatable(t, Clock)
end

function Clock.__tostring(t)
    return string.format("%02d:%02d", t.hours, t.minutes)
end

function Clock:plus(minutes)
    return Clock.at(self.hours, self.minutes + minutes)
end

function Clock:minus(minutes)
    return Clock.at(self.hours, self.minutes - minutes)
end

function Clock:equals(other)
    return self.hours == other.hours and self.minutes == other.minutes
end

return Clock
